### PR TITLE
Add std::io::{Read,Write} wrappers

### PR DESCRIPTION
**Pull request by fkrull**
See #160. I've only done the Read impl for now because figuring out the API was trickier than I expected, so I wanted to get some feedback first.

So it turns out you can't just implement Read for any type that implements InputStreamExt (or some other trait). I came up with a few ideas for that:

* Implement the trait for the InputStream struct, and all the structs for derived classes. Not great because it's not generic, and only works if you have the concrete type anyway.
* Implement Read on a helper struct and allow conversion into that helper using `From`.
* Implement Read on a helper struct and add an explicit method to convert to it.

I went with the last option because converting with `From` and `Into` always seemed to unsightly type annotations, involving the concrete helper type. With a custom method, the helper type can be more anonymous.

Interestingly, the compiler required the `InputStreamRead` struct to be public when using the function from a different crate, *but it didn't require the calling code to be able to actually see it*. So now client code can call `into_read` and use the return value, but it can't name its concrete type. Which feels a little weird.

### CODE REVIEW

File: src/input_stream.rs [hunk 1]
```diff
@@ -12,9 +12,11 @@ use glib::translate::*;
 use glib::Priority;
 use glib_ffi;
 use gobject_ffi;
+use std::io;
 use std::mem;
 use std::ptr;
 use InputStream;
+use super::error::to_std_io_result;
```
Comment (EPashkin) [inline src/input_stream.rs]: without `super::` it not works?

**Comment by EPashkin** [comment]
Thanks, this part looks good for me

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
```
Comment (GuillaumeGomez) [inline src/error.rs]: Please remove extra line.

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+use Error;
+use IOErrorEnum;
+use std::io;
+
```
Comment (GuillaumeGomez) [inline src/error.rs]: Please remove extra line.

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+impl From<IOErrorEnum> for io::ErrorKind {
+    fn from(kind: IOErrorEnum) -> Self {
+        match kind {
+            IOErrorEnum::Failed => io::ErrorKind::Other,
```
Comment (GuillaumeGomez) [inline src/error.rs]: Since you return `io::ErrorKind::Other` a lot of times, why not grouping all error returning it?

File: src/input_stream.rs [hunk 5]
```diff
@@ -204,9 +210,19 @@ impl<O: IsA<InputStream> + IsA<glib::Object> + Clone + 'static> InputStreamExtMa
     }
 }
 
+pub struct InputStreamRead<T: InputStreamExtManual>(T);
```
Comment (sdroege) [inline src/input_stream.rs]: Might be useful to also have a function to get back to the original type, and/or implement `Deref` for it

**Comment by fkrull** [comment]
Ok, I added the std::io::Write part as well. If there's anything else holding up this PR, let me know.

File: src/error.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use Error;
+use IOErrorEnum;
+use std::io;
+
+impl From<IOErrorEnum> for io::ErrorKind {
```
Comment (sdroege) [inline src/error.rs]: This is generally very useful, thanks :)

**Comment by sdroege** [comment]
Looks all good to me apart from the things already mentioned in the first comments above, which are mostly cosmetic :) Update that and we can get this in

**Comment by fkrull** [comment]
> 
> 
> Looks all good to me apart from the things already mentioned in the first comments above, which are mostly cosmetic :) Update that and we can get this in

Oh, I fixed those earlier, I didn't add a comment for it though. Or is there anything else I missed?

**Comment by sdroege** [comment]
> Oh, I fixed those earlier, I didn't add a comment for it though. Or is there anything else I missed?

Ah I just saw that the comments are still shown by github and it didn't detect them as "resolved". Will double check later :)

**Comment by sdroege** [comment]
Indeed, looks all good. @GuillaumeGomez ?

File: src/input_stream.rs (hunk 5 above)
Comment (sdroege) [reply src/input_stream.rs]: Should implement `Eq` and `Debug` here at least, based on the impls of the contained type. Can probably be derived

File: src/output_stream.rs [hunk 7]
```diff
@@ -179,9 +186,30 @@ impl<O: IsA<OutputStream> + IsA<glib::Object> + Clone + 'static> OutputStreamExt
     }
 }
 
+pub struct OutputStreamWrite<T: OutputStreamExt>(T);
```
Comment (sdroege) [inline src/output_stream.rs]: Should implement `Eq` and `Debug` here at least, based on the impls of the contained type. Can probably be derived

**Comment by GuillaumeGomez** [comment]
Looks good to me as well. Thanks!

**Comment by sdroege** [comment]
@fkrull can you send a new PR that implements `Debug` and `PartialEq`and `Eq` for the new structs :)

**Comment by sdroege** [comment]
And potentially something to come back from the new struct to the contained type again

**Pull request merged**